In [1]:
%reload_ext XXX.mumble

In [2]:
    get_ipython().ast_node_interactivity='last_expr_or_assign'

    get_ipython().ast_node_interactivity='last_expr_or_assign'

In [3]:
    from poser import *

    from poser import *

In [11]:
    ---
    urls:
        prefix: http://results.enr.clarityelections.com/GA/
        reports: http://results.enr.clarityelections.com/GA/63991/184321/en/reports.html
        ga: http://results.enr.clarityelections.com/GA/
        elections: http://results.enr.clarityelections.com/GA/elections.json
        county: http://results.enr.clarityelections.com/GA/63991/184321/en/select-county.html
        franklin: http://results.enr.clarityelections.com/GA/Franklin/64051/
        franklin_redir: http://results.enr.clarityelections.com/GA/Franklin/64051/183122/en/summary.html
        json: http://results.enr.clarityelections.com/GA/63991/184321/json/status.json
        select: http://results.enr.clarityelections.com/GA/63991/184321/en/select-county.html
        wiki: https://en.wikipedia.org/wiki/List_of_counties_in_Georgia_(U.S._state)

    ---
    urls:
        prefix: http://results.enr.clarityelections.com/GA/
        reports: http://results.enr.clarityelections.com/GA/63991/184321/en/reports.html
        ga: http://results.enr.clarityelections.com/GA/
        elections: http://results.enr.clarityelections.com/GA/elections.json
        county: http://results.enr.clarityelections.com/GA/63991/184321/en/select-county.html
        franklin: http://results.enr.clarityelections.com/GA/Franklin/64051/
        franklin_redir: http://results.enr.clarityelections.com/GA/Franklin/64051/183122/en/summary.html
        json: http://results.enr.clarityelections.com/GA/63991/184321/json/status.json
        select: http://results.enr.clarityelections.com/GA/63991/184321/en/select-county.html
        wiki: https://en.wikipedia.org/wiki/List_of_counties_in_Georgia_(U.S._state)

In [12]:
    read_html, get = (a/cache)((a.pandas.read_html(), a.requests.get()))
    num_counties = (a[read_html].drop(2).first().len()[x-1])(urls['wiki'])

    read_html, get = (a/cache)((a.pandas.read_html(), a.requests.get()))
    num_counties = (a[read_html].drop(2).first().len()[x-1])(urls['wiki'])

159

In [14]:
# A List of All of the Counties

    soup = a[get][x.text].bs4.BeautifulSoup(features='lxml')(urls['county'])
    county = (a.map(x.attrs['value']).list())(soup.select('table tr li a'));
    assert len(county) is num_counties, """The number of rows is not the number of counties."""

# A List of All of the Counties

    soup = a[get][x.text].bs4.BeautifulSoup(features='lxml')(urls['county'])
    county = (a.map(x.attrs['value']).list())(soup.select('table tr li a'));
    assert len(county) is num_counties, """The number of rows is not the number of counties."""

In [17]:
# The County Event ID

    event_ids = a.map(a.str.split('/').filter(bool).take(2)).dict()(county);
    a.dict.items().take(3).dict()(event_ids)

# The County Event ID

    event_ids = a.map(a.str.split('/').filter(bool).take(2)).dict()(county);
    a.dict.items().take(3).dict()(event_ids)

{'Appling': '63993', 'Atkinson': '63994', 'Bacon': '63995'}

In [19]:
    def construct_context_version(name='Appling', sleep=1):
        the.time.sleep(sleep)()
        latest_version = a[[a.identity(), event_ids.__getitem__]][
            [prefix], a.identity(), ['current_ver.txt']
        ].concat()['/'.join][get][x.text]

        constructed = '/'.join((prefix, name, event_ids[name], latest_version(name)))
        constructed = (constructed +'/json/{}.json').format
        return a.map(constructed).map((
            a.str.rsplit('/', 1).last().str.replace('.json', ''), 
            a[get].methodcaller('json'))).dict()(('sum', 'details', 'status'))

    def construct_context_version(name='Appling', sleep=1):
        the.time.sleep(sleep)()
        latest_version = a[[a.identity(), event_ids.__getitem__]][
            [prefix], a.identity(), ['current_ver.txt']
        ].concat()['/'.join][get][x.text]

        constructed = '/'.join((prefix, name, event_ids[name], latest_version(name)))
        constructed = (constructed +'/json/{}.json').format
        return a.map(constructed).map((
            a.str.rsplit('/', 1).last().str.replace('.json', ''), 
            a[get].methodcaller('json'))).dict()(('sum', 'details', 'status'))

In [25]:
    ---
    labels:
        CH: Candidate
        C: Office
        V: Votes
        TV: Total Voters
        BC: Ballot Cast

    ---
    labels:
        CH: Candidate
        C: Office
        V: Votes
        TV: Total Voters
        BC: Ballot Cast

In [49]:
    def tidy_precinct(s):
        return a.zip().dict().valmap(
            a.zip(s.Candidate).dict()
        ).pandas.DataFrame()(s.P, s.V)

    def tidy_precinct(s):
            return a.zip().dict().valmap(
                a.zip(s.Candidate).dict()
            ).pandas.DataFrame()(s.P, s.V)

In [ ]:
    dfs = {}
    for i, county in enumerate(event_ids):
        objects = construct_context_version(county)
        
        df = (a.pandas.DataFrame())(objects['sum']['Contests'])
        
        assert df.V.apply(sum).lt(df.TV).all()
        df = df[list(labels)].rename(columns=labels)
        df = (a.pandas.DataFrame()[df.join])(d['details']['Contests'])

        dfs[county] = a.pandas.concat()({index: tidy_precinct(row)
            for index, row in df.set_index('Office').iterrows() 
            if (a**list)(row.P) or not a.pandas.np.isnan()(row.P)})

    dfs = {}
    for i, county in enumerate(event_ids):
        objects = construct_context_version(county)
        
        df = (a.pandas.DataFrame())(objects['sum']['Contests'])
        
        assert df.V.apply(sum).lt(df.TV).all()
        df = df[list(labels)].rename(columns=labels)
        df = (a.pandas.DataFrame()[df.join])(d['details']['Contests'])

        dfs[county] = a.pandas.concat()({index: tidy_precinct(row)
            for index, row in df.set_index('Office').iterrows() 
            if (a**list)(row.P) or not a.pandas.np.isnan()(row.P)})

In [87]:
     a.pandas.concat()(dfs)

     a.pandas.concat()(dfs)

Attapulgus  \
Appling  CC, D2                                             T. WILKERSON (I) D                  NaN   
         CC, D4                                             R. RENTZ (I) R                      NaN   
         CC, D5                                             C. LEGGETT (I) R                  501.0   
         CLERK SUPERIOR COURT                               M. THOMAS (I) R                  1000.0   
         CORONER                                            J. MURPHY (I) R                    11.0   
         Constitutional Amendment #1<br>Provides greater... NO                                242.0   
                                                            YES                               833.0   
         Constitutional Amendment #2<br>Authorizes penal... NO                                366.0   
                                                            YES                               656.0   
         Constitutional Amendment #3<br>Reforms and re-e... NO                                258.0   
                                                            YES                               823.0   
         District Attorney, Brunswick Circuit               JACKIE JOHNSON (I) (REP)          965.0   
         President of the United States                     DONALD J. TRUMP (REP)             386.0   
                                                            GARY JOHNSON (LIB)                 11.0   
                                                            HILLARY CLINTON (DEM)             749.0   
         Public Service Commission, District 2 - Eastern    ERIC HOSKINS (LIB)                413.0   
                                                            TIM ECHOLS (I) (REP)              479.0   
         SHERIFF                                            M. W. MELTON (R)                  944.0   
         SURVEYOR                                           M. TOMBERLIN (I) R                965.0   
         State Representative, District 156                 GREG MORRIS (I) (REP)               NaN   
         State Representative, District 178                 CHAD NIMMER (I) (REP)             406.0   
         State Senator, District 19                         BLAKE TILLERY (REP)               718.0   
         TAX COMMISSIONER                                   D. CARTER (I) R                   268.0   
         U.S. Representative, District 12                   PATRICIA C. MCCRACKEN (DEM)       791.0   
                                                            RICK ALLEN (I) (REP)              325.0   
         United States Senator, Isakson                     ALLEN BUCKLEY (LIB)                33.0   
                                                            JIM BARKSDALE (DEM)               660.0   
                                                            JOHNNY ISAKSON (I) (REP)          358.0   
Atkinson BOE, D2                                            S. F. COOP (I) D                  833.0   
         BOE, D4                                            D. R. SMITH (R)                   656.0   
...                                                                                             ...   
Candler  U.S. Representative, District 12                   PATRICIA C. MCCRACKEN             791.0   
                                                            RICK ALLEN (I)                    325.0   
         United States Senator, Isakson                     ALLEN BUCKLEY                      33.0   
                                                            JIM BARKSDALE                     660.0   
                                                            JOHNNY ISAKSON (I)                358.0   
Carroll  CC D1                                              T. NORTH (I) D                    656.0   
         CC D3                                              T. LEE (I) R                      823.0   
         CHAIR COMMISSIONER                                 M. W. SMITH 

In [ ]:
    state_elections = a.pandas.DataFrame()(
        a.requests.get().methodcaller("json")(urls['elections'])
    ).set_index('EID')
    state_elections

> I want election **results** by **precinct** in **Georgia** for the **2016 presidential election**. The problem is that results are broken down by county in details.xls on the Georgia Secretary of State's website. However, precinct-level details are available for each of Georgia's 159 counties 

![](./state_list.png)

in individual reports (also titled details.xls). 

![](./county_example.png)